langchain/retrievers/multi_query.py

```diff
@@ -0,0 +1,146 @@
+
+        Returns:
+            MultiQueryRetriever
+        """
+        self.retriever = retriever
+        self.num_queries = num_queries
+        self.llm_chain = llm_chain
+        self.verbose = verbose
+
+    @classmethod
+    def from_llm(
+        cls, retriever: BaseRetriever, num_queries: int, llm: BaseLLM
+    ) -> "MultiQueryRetriever":
+        """Initialize from llm using default template.
+
+        Args:
+            retriever: retriever to query documents from
+            num_queries: number of queries for the LLM to generate
+            llm: llm for query generation using DEFAULT_QUERY_PROMPT
+
+        Returns:
+            MultiQueryRetriever
+        """
+        llm_chain = LLMChain(llm=llm, prompt=cls.DEFAULT_QUERY_PROMPT)
+        return cls(retriever=retriever, num_queries=num_queries, llm_chain=llm_chain)
+
+    def get_relevant_documents(self, question: str) -> List[Document]:
+        """Get relevated documents given a user query.
+
+        Args:
+            question: user query
+
+        Returns:
+            Unique union of relevant documents from all generated queries
+        """
+        queries = self.generate_queries(question)
+        documents = self.retrieve_documents(queries)
+        unique_documents = self.unique_union(documents)
+        return unique_documents
+
+    def generate_queries(self, question: str) -> List[str]:
+        """Generate queries based upon user input.
+
+        Args:
+            question: user query
+
+        Returns:
+            List of LLM generated queries that are similar to the user input
+        """
+        response = self.llm_chain.apply(
+            [{"question": question, "num_queries": self.num_queries}]
+        )
+        # Get the response text
+        response_text = response[0]["text"]
+        # Parse the response
+        output_parser = LineListOutputParser()
+        parsed_output = output_parser.parse(response_text)
+        if self.verbose:
+            print(f"Generated queries: {parsed_output.lines}")
+        return parsed_output.lines
+
+    def retrieve_documents(self, queries: List[str]) -> List[Document]:
+        """Run all LLM generated queries.
+
+        Args:
+            queries: query list
+
+        Returns:
+            List of retrived Documents
+        """
+        documents = []
+        for query in queries:
+            docs = self.retriever.get_relevant_documents(
+                query
+            )  # assuming this method exists
+            documents.extend(docs)
+        return documents
+
+    def unique_union(self, documents: List[Document]) -> List[Document]:
+        """Get uniqe Documents.
+
+        Args:
+            documents: List of retrived Documents
+
+        Returns:
+            List of unique retrived Documents
+        """
+        # Create a dictionary with page_content as keys to remove duplicates
+        unique_documents_dict = {doc.page_content: doc for doc in documents}
```

Comment:
this doesn't account for metadata right? we would probably want to account for that

Reply:
Hmm, is the retrieved text (`page_content`) is identical I assume we would want to collapse into a single doc even if metadata is slightly different?

Reply:
metadata can be important, i would probably hash that as well

Reply:
Metadata can be misleading. For example, we have the same `document_content` requested from different sources. If we store the `source_url` as metadata, then we get different keys for the same content. 
Maybe, we can include metadata in the key with more control? For example, add the `unique_metadata_keys` parameter. 
Maybe it is overengineering and content as the key is enough?

Reply:
yeah but if that source_url is important later on, we're introducing some noise

lets make parameterizable at very least

i would default to including metadata in hash

Reply:
got it. will add. @leo-gan thanks for the feedback, too.